core/src/main/scala/scala/pickling/pickler/Either.scala

```diff
@@ -1,69 +1,180 @@
         builder.putField("b", b => rp.pickle(picklee.b, b))
         builder.endEntry()
       }
+
       override def unpickle(tag: String, reader: PReader): Any = {
         if (t.key == tag) {
           val rr = reader.readField("b")
-          if(rp.tag.isEffectivelyPrimitive) rr.hintElidedType(rp.tag)
+          if (rp.tag.isEffectivelyPrimitive) rr.hintElidedType(rp.tag)
           Right(ru.unpickleEntry(rr).asInstanceOf[R])
-        } else throw new PicklingException(s"RightUnpickler can't unpickle: $tag")
+        } else throw TypeMismatch(List(t), List(FastTypeTag(tag)))
       }
+
       override def toString = s"RightPicklerUnpickler($tag)"
     }
+  }
 
-  implicit def pickleUnpickleEither[L,R](implicit rp: Pickler[Right[L,R]], ru: Unpickler[Right[L, R]],
-                                         lp: Pickler[Left[L,R]], lu: Unpickler[Left[L, R]],
-                                         t: FastTypeTag[Either[L,R]]): AbstractPicklerUnpickler[Either[L,R]] =
+  implicit def pickleUnpickleEither[L,R]
+    (implicit rp: Pickler[Right[L,R]], ru: Unpickler[Right[L, R]],
+              lp: Pickler[Left[L,R]], lu: Unpickler[Left[L, R]],
+              t: FastTypeTag[Either[L,R]]): AbstractPicklerUnpickler[Either[L,R]] = {
     new AbstractPicklerUnpickler[Either[L, R]] with AutoRegister[Either[L, R]] {
       override def pickle(picklee: Either[L, R], builder: PBuilder): Unit = {
         picklee match {
-          case l: Left[L,R] => lp.pickle(l, builder)
-          case r: Right[L,R] => rp.pickle(r, builder)
+          case l: Left[L, R] => lp.pickle(l, builder)
+          case r: Right[L, R] => rp.pickle(r, builder)
         }
       }
+
       override def unpickle(tag: String, reader: PReader): Any = {
-        if(tag == rp.tag.key) ru.unpickle(tag,reader)
-        else if(tag == lp.tag.key) lu.unpickle(tag, reader)
-        else throw new PicklingException(s"Unknown type tag for Either: $tag")
+        if (tag == rp.tag.key) ru.unpickle(tag, reader)
+        else if (tag == lp.tag.key) lu.unpickle(tag, reader)
+        else throw TypeMismatch(List(rp.tag, lp.tag), List(FastTypeTag(tag)))
       }
+
       override def tag: FastTypeTag[Either[L, R]] = t
-      override def toString = s"EitherPicklerUnpickler($t)"
+
+      override def toString = s"EitherPicklerUnpickler($tag)"
     }
+  }
+
+  locally {
+
+    registerPicklerAsGen(RuntimeLeftPicklerUnpickler)
+    registerPicklerAsGen(RuntimeRightPicklerUnpickler)
+    registerPicklerAsGen(RuntimeEitherPicklerUnpickler)
+
+  }
+
+}
+
+trait EitherPicklersRuntime extends GeneratorHelper {
+
+  /** Pickle as [[Any]] because the scala pickling
+    * use [[Any]] as a placeholder for the type params.
+    */
+  private def pickleAsAny[S <: Either[_, _]]
+      (picklee: S, fullTpe: FastTypeTag[S],
+       name: String, field: Any, builder: PBuilder) = {
+
+    builder.beginEntry(picklee, fullTpe)
+    builder.putField(name, { b =>
+      AnyPicklerUnpickler.pickle(field, b)
```

Comment:
Just a note, it would be feasible to actually NOT force runtime either to use Any given the way type tags work.  i.e. you could generate the Either pickler with concrete tags.

This is probably fine as-is though :)